src/dataflow/render/reduce.rs

```diff
@@ -134,15 +132,29 @@ where
                 // in differential dataflow.
                 let oks = differential_dataflow::collection::concatenate(scope, ok_partials)
                     .reduce_abelian::<_, OrdValSpine<_, _, _, _>>("ReduceCollation", {
+                    let aggregates_clone = aggregates.clone();
                     let aggregates_len = aggregates.len();
                     move |key, input, output| {
                         // The intent, unless things are terribly wrong, is that `input`
-                        // contains, in order, the values to drop into `output`.
-                        assert_eq!(input.len(), aggregates_len);
+                        // contains, in order, the values to drop into `output`. If this
+                        // is not the case, we should express our specific discontent.
+                        if input.len() != aggregates_len || input.iter().enumerate().any(|(i,((p,_),_))| &i != p) {
+                            // TODO(frank): Arguably, the absence of one aggregate is evidence
+                            // that the key doesn't exist (the others could be phantoms due to
+                            // negative input records); we could just suppress the output in that
+                            // case, rather than panic, though we surely want to see what is up.
+                            // XXX: This panic reports user-supplied data!
+                            panic!(
```

Comment:
I think at this point we only have the input relation expr, though I guess that is enough. My plan was to next tweak the RTE code to put the operator information in with the errors, which would allow one to start things up again and look at the operator identifiers. We could certainly stash a copy of the whole relation expr to dump here, though.

Reply:
Dumped the relation expr as part of the panic. Ideally we'll walk some of this back when we have more confidence that it doesn't error, but ... yeah.